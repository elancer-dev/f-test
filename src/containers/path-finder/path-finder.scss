html,
body {
    width: 100%;
    height: 100%;
    margin: 0;
    padding: 0;
    font-family: Arial, Helvetica, sans-serif;
}

.path-finder {
    width: 100vw;
    height: 100vh;
    display: flex;
}

.path-finder__sidebar {
    width: 300px;
    flex-shrink: 0;
}

.path-finder__map-container {
    flex-grow: 1;
}

.path-finder__map {
    width: 100%;
    height: 100%;
}

.path-finder__searct-input {
    display: block;
    margin: 20px 10px;
    width: 276px;
}