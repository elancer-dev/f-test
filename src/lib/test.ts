import { TState } from "../utils/types";

export const getState = (): TState => {
    return {
        points: [
            { coord: [0, 0], name: 'name1' },
            { coord: [1, 1], name: 'name2' },
            { coord: [2, 2], name: 'name3' },
            { coord: [3, 3], name: 'name4' },
        ]
    }
}