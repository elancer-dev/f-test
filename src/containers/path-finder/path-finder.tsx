import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { TState, TActionFunction, TPoint, YMapsEvent } from './../../utils/types';
import { action, actionsHandler } from './../../utils/actions';
import { YMaps, Map, YMapsApi } from 'react-yandex-maps';
import PointsList from './../points-list/points-list';
import './path-finder.scss';

const Boundary: React.FC = ({ children }: { children?: React.ReactNode; }) => (<Fragment>{process.env.NODE_ENV === 'test' ? null : children}</Fragment>);

type TProps = {}

type TPS = TProps & {
    points: Array<TPoint>;
    action: TActionFunction;
}

export class PathFinder extends React.PureComponent<TPS> {

    private ymaps: YMapsApi | null = null;
    private map: any = null;
    private route: any = null;
    private balloonLayout: any = null;
    private suggestInput: React.RefObject<HTMLInputElement>;

    constructor(props: TPS) {
        super(props);
        this.suggestInput = React.createRef();
    }

    mapOnLoadHolder = (ymaps: YMapsApi) => {

        this.ymaps = ymaps;

        ymaps.load((ymaps: YMapsApi) => {

            const suggestView = new ymaps.SuggestView("suggest", {
                provider: 'yandex#map',
                results: 10,
            });

            suggestView.events.add('select', (e: YMapsEvent) => {

                if (this.suggestInput.current) {
                    this.suggestInput.current.value = '';
                }

                var item = e.get('item');

                ymaps.geocode(item.value).then((res: any) => {

                    var coord = res.geoObjects.get(0).geometry.getCoordinates();

                    this.props.action(actionsHandler, 'addPoint', { coord: coord, name: item.displayName });

                });

            }, this);

            this.balloonLayout = ymaps.templateLayoutFactory.createClass('<div>{{ properties.address|raw }}</div>');

        });

    }

    routeAppend = () => {

        if (this.ymaps) {

            var multiRoute = new this.ymaps.multiRouter.MultiRoute({ referencePoints: this.props.points.map(point => point.name) },
                {
                    editorMidPointsType: "way",
                    editorDrawOver: false,
                    preventDragUpdate: true,
                    balloonContentLayout: this.balloonLayout,
                    balloonPanelMaxMapArea: 0,
                    boundsAutoApply: true,
                });

            multiRoute.editor.start({
                addWayPoints: false,
                dragWayPoints: true,
                removeWayPoints: false,
                addMidPoints: false,
            });

            if (this.route) {
                this.map.geoObjects.remove(this.route);
            }

            this.route = multiRoute;
            this.map.geoObjects.add(multiRoute);

            multiRoute.model.events.once("requestsuccess", () => {

                var points = multiRoute.getWayPoints();

                for (var i = 0; i < points.getLength(); i++) {

                    var point = points.get(i);

                    if (this.ymaps) {
                        this.ymaps.geoObject.addon.balloon.get(point);
                    }

                    point.options.set({
                        preset: "islands#blueIcon",
                        iconContentLayout: this.ymaps ? this.ymaps.templateLayoutFactory.createClass((i + 1).toString()) : '',
                        balloonContentLayout: this.balloonLayout,
                        balloonOffset: [0, -35],
                        hideIconOnBalloonOpen: false,
                        draggable: true,
                        pointIndex: i,
                    });

                }

            });

            multiRoute.editor.events.add('waypointdragend', (e: any) => {

                var point = e.get('wayPoint');
                var coord = point.geometry.getCoordinates();
                var pointIndex = point.options.get('pointIndex');

                if (this.ymaps) {

                    this.ymaps.geocode(coord, { results: 1 }).then((res: any) => {

                        var properties = res.geoObjects.get(0).properties.getAll();

                        this.props.action(actionsHandler, 'updatePoint', { index: pointIndex, coord: coord, name: properties.text });

                    });
                }

            });

        }

    }

    render = () => {

        this.routeAppend();

        return (

            <div className="path-finder" >
                <div className="path-finder__sidebar">
                    <input type="text" autoFocus={true} id="suggest" className="path-finder__searct-input" ref={this.suggestInput} />
                    <Boundary>
                        <PointsList />
                    </Boundary>
                </div>
                <div className="path-finder__map-container">
                    <Boundary>
                        <YMaps
                            query={{
                                ns: "use-load-option",
                                apikey: "3f58b004-e8ca-4980-bdfb-03221ab45a61",
                                load: "Map"
                            }}>
                            <Map
                                className="path-finder__map"
                                defaultState={{ center: [55.75, 37.57], zoom: 9, }}
                                instanceRef={(map: React.Ref<YMapsApi>) => { this.map = map }}
                                onLoad={this.mapOnLoadHolder} />
                        </YMaps>
                    </Boundary>
                </div>
            </div>

        )

    }

}

const mapStateToProps = (state: TState) => {

    return {
        points: state.points,
    };

};

const mapDispatchToProps = (dispatch: Dispatch) => {

    return bindActionCreators({ action: action }, dispatch);

};

export default connect(mapStateToProps, mapDispatchToProps)(PathFinder);