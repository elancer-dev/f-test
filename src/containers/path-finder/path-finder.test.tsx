import React from 'react';
import TestRenderer from 'react-test-renderer';
import { getState } from "./../../lib/test";
import { PathFinder } from './path-finder';

export default describe('<PathFinder />', () => {

    it('Renders PathFinder component', () => {

        const action = jest.fn();
        const state = getState();

        const root = <PathFinder action={action} points={state.points} />;

        const tree = TestRenderer.create(root).toJSON();

        expect(tree).toMatchSnapshot();

    });

});