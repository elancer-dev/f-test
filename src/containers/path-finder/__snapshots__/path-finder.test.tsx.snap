// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PathFinder /> Renders PathFinder component 1`] = `
<div
  className="path-finder"
>
  <div
    className="path-finder__sidebar"
  >
    <input
      autoFocus={true}
      className="path-finder__searct-input"
      id="suggest"
      type="text"
    />
  </div>
  <div
    className="path-finder__map-container"
  />
</div>
`;
